Show plugin release date in plugin manager

daniel-beck opened the pull request:
This was one of the feedbacks at the contributor summit at FOSDEM: It would be helpful to see at a glance when a plugin was released, to help understand how actively it is maintained.

Up to date screenshots in https://github.com/jenkinsci/jenkins/pull/4535#issuecomment-593385560

### Proposed changelog entries

* Show plugin release date in plugin manager

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [n/a] JIRA issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [n/a] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

daniel-beck committed:
Show plugin release date in plugin manager

oleg-nenashev commented:
One comment about UX, label display names should probably come from the update center: https://github.com/jenkins-infra/plugin-site-api/issues/82 .Though it keeps a question about custom update centers then

StefanSpieker reviewed (approved):
That is a nice feature. Thanks 👍

timja commented:
Alignment of the version column seems wrong on the available tab:

![image](https://user-images.githubusercontent.com/21194782/75623098-d2557e00-5b9e-11ea-96d0-072305810e78.png)

It's fine on the installed column, and I don't have any plugin updates atm, but I assume it's fine there as that's what is screenshotted

I was also happy to see that the date is localised and I get the date format that I am used to

timja reviewed (dismissed):
See https://github.com/jenkinsci/jenkins/pull/4535#issuecomment-593075067,

should be a minor fix hopefully

timja commented:
Hmm taking another look without this change,
it already seems to be misaligned but this makes it worse,

Removing my change requested but I still think it should be improved:

![image](https://user-images.githubusercontent.com/21194782/75623158-6293c300-5b9f-11ea-877e-8d19690891d5.png)

daniel-beck commented:
> Alignment of the version column seems wrong on the available tab:

It's due to ridiculously long version numbers like `1.12+build.201809061734` or `3.8.20200204-101107.d1d344b` and left-alignment of the column content. Unrelated to this change.

oleg-nenashev commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -1018,6 +1020,12 @@ private static String get(JSONObject o, String prop) {
         private Set<Plugin> incompatibleParentPlugins;
 
+        /**
+         * Date when this plugin was released.
```
```suggestion
         * Date when this plugin was released.
         * @since TODO
```

oleg-nenashev reviewed (approved):
I wonder whether formatting would differ depending on the user locale, but it is fine even if it is not

daniel-beck commented:
> I wonder whether formatting would differ depending on the user locale

It does, that's the whole point of using this control.

daniel-beck committed:
Add at-since

Co-Authored-By: Oleg Nenashev <o.v.nenashev@gmail.com>

timja commented:
> I wonder whether formatting would differ depending on the user locale, but it is fine even if it is not

if you see my screenshot it has a different locale to DB's

daniel-beck committed:
Make sorting work property

- Sort independently of sortable.js date parsing magic
- Hide categories when sorting by release date

daniel-beck committed:
Merge branch 'plugin-release-date' of github.com:daniel-beck/jenkins into plugin-release-date

daniel-beck commented:
This now depends on https://github.com/jenkinsci/jenkins/pull/4534 not being merged before this, or https://github.com/jenkinsci/jenkins/pull/4535/files#diff-6c89c5cced7eb382005a0b0825a45140R70 needs `onclick` removed otherwise. Should be trivial to do on GitHub and ideally not hold off merging for technicalities.

fqueiruga commented:
I know I'm late to the party but I have a suggestion. Can we consider, instead of having another numeric column showing the date, showing the last release as  text? That will remove mental load for the user.

We could also show that message within the version column. The information relates to the version anyway. An example:
<img width="778" alt="Captura de pantalla 2020-03-02 a las 10 24 53" src="https://user-images.githubusercontent.com/5738588/75662923-3c8a2380-5c70-11ea-9587-388874663fc7.png">

daniel-beck commented:
> We could also show that message within the version column. The information relates to the version anyway.

I considered that, but was unsure how this would impact sorting by this column. Given how useless sort by version is, I guess this column would then sort by release date? Is that useful UX?

fqueiruga commented:
If we want to sort by release date, we would definitely need a column just for it IMO

daniel-beck commented:
> If we want to sort by release date, we would definitely need a column just for it IMO

Well, the table is sortable. Right now it's possible to sort by name and version number (🤣). Sorting by date is potentially the most useful of any of them (even if there's no obvious use case other than something like getting general ecosystem status at a glance).

Anyway, I like your "time span" idea a lot better than just displaying the date. I will implement that. The specific date can always be the tool tip.

fqueiruga commented:
May I also suggest using the `<time>` tag? With tooltip would look something like this `<time datetime="2019-08-15" tooltip="15/08/2019">8 months ago</time>`.

daniel-beck committed:
Show friendlier labels for the age of plugin releases

daniel-beck commented:
Updated based on feedback. It now looks like this:

![Screenshot 2020-03-02 at 13 34 16](https://user-images.githubusercontent.com/1831569/75677822-98fb3c00-5c8c-11ea-8477-d7e7ef499e81.png)
![Screenshot 2020-03-02 at 13 37 17](https://user-images.githubusercontent.com/1831569/75677828-9b5d9600-5c8c-11ea-8ef4-5d7af0bdb68b.png)

jvz reviewed (approved):
Looks nice!

daniel-beck commented:
Universal approval, so marking `ready-for-merge` so this can go into the upcoming weekly.

timja commented:
@daniel-beck are you able to resolve the conflicts please?

daniel-beck committed:
Merge branch 'master' into plugin-release-date

daniel-beck commented:
\*cough\* https://github.com/jenkinsci/jenkins/pull/4535#issuecomment-593150643 \*cough\* 😁

timja commented:
Lol, other one had been updated more recently so it was higher in the list, and your comment got buried above many other comments :p
